lib/msf/core/db_manager/import/masscan.rb

```diff
@@ -0,0 +1,259 @@
+          if fix_services
+            port_states = h["ports"].map {|p| p["state"]}.reject {|p| p == "filtered"}
+            next if port_states.compact.empty?
+          end
+          yield(:address,data[:host]) if block
+          hobj = report_host(data)
+          report_import_note(wspace,hobj)
+        end
+      end
+
+      if( h["os_vendor"] )
+        note = {
+          :workspace => wspace,
+          :host => hobj || addr,
+          :type => 'host.os.masscan_fingerprint',
+          :task => args[:task],
+          :data => {
+            :os_vendor   => h["os_vendor"],
+            :os_family   => h["os_family"],
+            :os_version  => h["os_version"],
+            :os_accuracy => h["os_accuracy"]
+          }
+        }
+
+        if(h["os_match"])
+          note[:data][:os_match] = h['os_match']
+        end
+
+        report_note(note)
+      end
+
+      if (h["last_boot"])
+        report_note(
+          :workspace => wspace,
+          :host => hobj || addr,
+          :type => 'host.last_boot',
+          :task => args[:task],
+          :data => {
+            :time => h["last_boot"]
+          }
+        )
+      end
+
+      if (h["trace"])
+        hops = []
+        h["trace"]["hops"].each do |hop|
+          hops << {
+            "ttl"     => hop["ttl"].to_i,
+            "address" => hop["ipaddr"].to_s,
+            "rtt"     => hop["rtt"].to_f,
+            "name"    => hop["host"].to_s
+          }
+        end
+        report_note(
+          :workspace => wspace,
+          :host => hobj || addr,
+          :type => 'host.masscan.traceroute',
+          :task => args[:task],
+          :data => {
+            'port'  => h["trace"]["port"].to_i,
+            'proto' => h["trace"]["proto"].to_s,
+            'hops'  => hops
+          }
+        )
+      end
+
+
+      # Put all the ports, regardless of state, into the db.
+      h["ports"].each { |p|
+        # Localhost port results are pretty unreliable -- if it's
+        # unknown, it's no good (possibly Windows-only)
+        if (
+          p["state"] == "unknown" &&
+          h["status_reason"] == "localhost-response"
+        )
+          next
+        end
+        extra = ""
+        extra << p["product"]   + " " if p["product"]
+        extra << p["version"]   + " " if p["version"]
+        extra << p["extrainfo"] + " " if p["extrainfo"]
+
+        data = {}
+        data[:workspace] = wspace
+        if fix_services
+          data[:proto] = masscan_msf_service_map(p["protocol"])
+        else
+          data[:proto] = p["protocol"].downcase
+        end
+        data[:port]  = p["portid"].to_i
+        data[:state] = p["state"]
+        data[:host]  = hobj || addr
+        data[:info]  = extra if not extra.empty?
+        data[:task]  = args[:task]
+        data[:name]  = p['tunnel'] ? "#{p['tunnel']}/#{p['name'] || 'unknown'}" : p['name']
+        report_service(data)
+      }
+      #Parse the scripts output
```

Comment:
This all seems copied from the nmap code and not necessary or relevant to masscan.